### PR TITLE
Refactor logout + Search bug

### PR DESCRIPTION
**Pull request by oliverheywood451**
## Description

There are two main changes here, a refactor to logout, and a bug fix around searching.

- We had three logout functions, now there is one in AppAuthService. We also had two ways of checking if a user was anonymous, the BehaviorSubject in AppStateService and a small function in appAuthService that looked directly at the cookies. Now only the BehaviorSubject is publicly available. Also IsLoggedIn was renamed to IsProfiledUser.

- The search bug came from a line of code that checked if the current route was "/products". It was affecting every instance of the search component.  

## Type of change

- [x ] Bug fix (non-breaking change which fixes an issue)
- [x ] Refactor (change/optimize code without changing external behavior)

## Checklist:

- [ x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x ] I have added tests that prove my fix is effective or that my feature works
- [x ] New and existing unit tests pass locally with my changes
- [ ] I have updated the acceptance criteria on the task so that it can be tested
- [ x] I have verified my contribution works and looks well in Firefox, Safari and Internet Explorer

### CODE REVIEW

File: src/UI/Buyer/src/app/shared/components/search/search.component.ts [hunk 1]
```diff
@@ -56,27 +49,21 @@ export class SearchComponent implements OnInit, OnDestroy {
...
   }
 
   clear(): void {
+    this.form.markAsDirty();
+    this.form.setValue({ search: '' });
+  }
+
+  clearWithoutEmit(): void {
     this.form.setValue({ search: '' });
   }
```
Comment (crhistianramirez) [inline src/UI/Buyer/src/app/shared/components/search/search.component.ts]: Doesn't look like clearWithoutEmit is being used?

File: src/UI/Buyer/src/app/layout/header/header.component.ts [hunk 2]
```diff
@@ -89,12 +92,21 @@ export class HeaderComponent implements OnInit, OnDestroy {
...
+    this.activatedRoute.queryParams
+      .pipe(
+        filter((queryParams) => {
+          return typeof queryParams.search === 'undefined';
+        }),
+        takeWhile(() => this.alive)
+      )
+      .subscribe(() => {
+        const search = this.isMobile() ? this.mobileSearch : this.desktopSearch;
+        if (search) search.clearWithoutEmit();
```
Comment (crhistianramirez) [inline src/UI/Buyer/src/app/layout/header/header.component.ts]: Can you wrap `search.clearWithoutEmit()` in brackets

Comment (oliverheywood451) [reply src/UI/Buyer/src/app/layout/header/header.component.ts]: will do.

File: src/UI/Buyer/src/app/shared/components/search/search.component.ts (hunk 1 above)
Comment (oliverheywood451) [reply src/UI/Buyer/src/app/shared/components/search/search.component.ts]: Its used in the other code you commented on ;)

Comment (crhistianramirez) [reply src/UI/Buyer/src/app/shared/components/search/search.component.ts]: lol facepalm

**Pull request merged**